index.js

```diff
@@ -189,7 +189,7 @@ module.exports = function (opts) {
           }
           this._expiryTimeout = setTimeout(function () {
             this.emit('expiry', token);
-          }.bind(this), new Date(token.expiry) - new Date());
+          }.bind(this), new Date(token.expiry) - new Date() - 1000);
```

Comment:
Do you think re-requesting st2 auth token `1s` before its expiry would be enough to cover real world communication overhead or we should lift it to something safer like for example `3s`?

cc @nmaludy

Reply:
Two ways I've seen this done previously:

1. Refresh every expiration_interval_sec/2. This way there is plenty of time to always get auth and timing issues are minimized. If the refresh interval is 24h, then you will get a new token every 12.

2. Refresh at a configurable offset: experation_interval_sec - configurable_offset_sec. This way if your with environment is loaded or slow, you can tune to ensure you have more headroom if you run into auth timing issues.

Reply:
For this specific timeout, I think the exact number is going to depend on the specific environment, so ideally we would make this a parameter or environment variable so users could set their own acceptable token refresh timeout.

But that might be overengineering it. Really, all we need to do is make sure that it requests a new token "soon enough" for the vast majority of use cases. And for that, I would take a look at the timeout values for other software.

According to [this StackOverflow post](https://stackoverflow.com/a/13582705), the default timeout settings for Firefox are:
* 115s for TCP TTL
* 250s for connection timeout
* 30s for HTTP request timeout

Further down in the comments somebody mentions that the [default timeout value for Apache is 300s](https://stackoverflow.com/questions/13582409/http-client-timeout-and-server-timeout#comment30120626_13582705), but how exactly this timeout is measured is not specified (and I didn't dig into it, since we don't use Apache).

The nginx timeout defaults are another source of information. According to [their documentation](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout), the default value for `proxy_read_timeout` is 60s, and that is for the timeout between "successive read operations".

So I would think ~35s until expiry is a good value for the timeout. That should pretty guarantee at least one full request/response cycle, and if we pad that number it would give some extra time for the server to generate a new token on top of the HTTP request timeout:

```
---- | ---------------------------------------------------- time (seconds) --> ------------------------------------------------ | -->
  refresh                                                                                                                old token expiry
---- | ----------------------------------------------------- ~35s "delta" ----------------------------------------------------- |
---- | -- request sent (15s) -- | -- server generates new token (~3s) -- | -- response sent (15s) -- | -- "slack time" (~2s) -- |
----------------------------------------- old token used ------------------------------------------- | ---- new token used --------->
```

(I have no idea how long it takes a server to generate a new token, I'm just putting in a number to illustrate my point)

That puts a lower bound on the expiry time of ~35s (eg: must be greater than the timeout value), but I'm not sure what value there would be for users to specify an expiry time of less than 35s to begin with.

What I would like to avoid is a situation where we continuously bump this delta value up by a little bit to satisfy users. If we start out with a timeout of 1s before expiry (and release a new version of st2client.js, and another version of hubot-stackstorm, and another version of st2chatops package), and then bump it up to 3s before expiry a few days later (and release a new version of st2client.js, and another version of hubot-stackstorm, and another version of st2chatops package), and then bump it up to 5s before expiry a few days after that (and release a new version of st2client.js, and another version of hubot-stackstorm, and another version of st2chatops package).

I would rather set the delta to a high value to start out with that gives a near guarantee that the token will be refreshed before the old token expires.

But I'm happy to have my mind changed if somebody has other ideas.

Reply:
Didn't see nmaludy's post before submitting my own comment.

Since st2client.js isn't 100% specific to ST2, I think setting the delta to the default expiration_interval_sec/2 is a little too specific to ST2. Different users will likely have different values for the expiration interval.

I think the second option, a "configurable offset" (I call it a "delta") in my comment is the way to go here.

And like I said, I don't want to continuously tune this variable in st2client.js and subsequently release new versions of st2client.js, hubot-stackstorm, and st2chatops. So I would push for a larger delta over a smaller delta.

Reply:
Yeah, from @nmaludy 1'st case `/2` looks pretty huge change since this piece of code should really emit event on `expiry` as the lib could be used in more places.

For 2'nd example from @nmaludy  we're not going to provide this option configurable but instead find something that works for 90% of cases.

I'm fine if it won't work if remote server is really slow and couldn't respond in `30s`, - that will escalate into other issues and not just token issues.

Just a note that `1s` diff, even `100ms` diff fixes race condition in majority of the cases when token _already_ expired.

Reply:
I think that's fair, and I would default to a higher value as blag suggested

Reply:
Thanks, I'm setting it to `10s` which is way higher to `1s` and should cover 90% cases.

If we'll need to lift it even more, - then something terribly wrong happens on user's side.

Reply:
@nmaludy @blag Thanks for input :+1: 

Changed to `10s` https://github.com/StackStorm/st2client.js/pull/69/commits/3a36fe32ae665fec4939649107125bd15bd4f1ec as a middle ground, per feedback.